src/app/core/services/notifications/message-handler.service.ts

```diff
@@ -0,0 +1,57 @@
+import { Injectable } from '@angular/core'
+import { FirebaseX } from '@ionic-native/firebase-x/ngx'
+
+import { UsageEventType } from '../../../shared/enums/events'
+import { Assessment, AssessmentType } from '../../../shared/models/assessment'
+import { MessagingAction } from '../../../shared/models/notification-handler'
+import { QuestionnaireService } from '../config/questionnaire.service'
+import { LogService } from '../misc/log.service'
+import { ScheduleService } from '../schedule/schedule.service'
+import { UsageService } from '../usage/usage.service'
+
+@Injectable()
+export class MessageHandlerService {
+  constructor(
+    public firebase: FirebaseX,
+    public logger: LogService,
+    public schedule: ScheduleService,
+    public questionnaire: QuestionnaireService,
+    public usage: UsageService
+  ) {
+    this.firebase
+      .onMessageReceived()
+      .subscribe(data => this.onMessageReceived(new Map(Object.entries(data))))
+  }
+
+  onMessageReceived(data) {
+    const action = data.get('action')
+    switch (action) {
+      case MessagingAction.QUESTIONNAIRE_TRIGGERED:
+        this.logger.log('A questionnaire was triggered!')
+        const questionnaire = <Assessment>JSON.parse(data.get('questionnaire'))
+        const metadata = data.get('metadata')
+        return this.triggerQuestionnaire(questionnaire).then(() =>
+          this.usage.sendQuestionnaireEvent(
+            UsageEventType.QUESTIONNAIRE_TRIGGERED,
+            questionnaire.name,
+            Date.now(),
+            metadata
+          )
+        )
+      default:
+        this.logger.log('Cannot process message.')
+    }
+  }
+
+  triggerQuestionnaire(questionnaire: Assessment) {
+    return this.questionnaire
+      .pullDefinitionForSingleQuestionnaire(questionnaire)
+      .then(() =>
+        this.schedule.generateSingleAssessmentTask(
+          questionnaire,
+          AssessmentType.SCHEDULED,
+          Date.now()
+        )
```

Comment:
will this send a notification for this questionnaire too?

I think the data message should display a notification on the device too? Otherwise, we can add it to the appserver (so data message + notification message).

Reply:
Do you mean the data message trigger itself? If yes, if we add the notification message/notification key then it should show the notification. In the recording here: https://i.ibb.co/nRf6RJ3/ezgif-6-5315e7dd916c.gif , I set the notification payload as well to show the notification. Also, do we want it to show even if the app is in the foreground?
```
{
  "notification": {
    "body": "Notification body",
    "title": "Notification title"
  },
  "data": {
    "notification_foreground": "true",
  }
}```

Or do you mean for example if we trigger a questionnaire for later? I haven't added an additional reschedule yet.

Reply:
No this seems fine, and yes we want to display the notification even if the app is in the foreground as this is something that they should not miss.
btw i am not able to access the recording you posted, probably because i am looking at it after so many days?

Reply:
Yeah I think it got auto deleted. Heres the new url: https://i.ibb.co/kH4HZLt/ezgif-6-5315e7dd916c.gif

Reply:
ok, thanks. yes looks good.